Migrate EcalClusterLazyTools to use ESGetToken

guitargeek opened the pull request:
#### PR description:

Migrate EcalClusterLazyTools to use ESGetToken. After https://github.com/cms-sw/cmssw/pull/31700, this is the next step to migrate the `RecoEgamma/EgammaElectronProducers` package, but it also affects many other packages.

#### PR validation:

CMSSW compiles, local matrix tests pass.

guitargeek committed:
const interface for EcalClusterLazyTools

guitargeek committed:
Migrate EcalClusterLazyTools to ESGetToken

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

ERROR: Build errors found during clang-tidy run.
```
RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:523:26: error: use of undeclared identifier 'egamma' [clang-diagnostic-error]
      TowersBehindClus = egamma::towersOf(*scRef, *ctmaph);
                         ^
RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:524:30: error: use of undeclared identifier 'egamma' [clang-diagnostic-error]
      hcalDepth1OverEcalBc = egamma::depth1HcalESum(TowersBehindClus, *hcalTowers) / scRef->energy();
                             ^
RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:525:30: error: use of undeclared identifier 'egamma' [clang-diagnostic-error]
      hcalDepth2OverEcalBc = egamma::depth2HcalESum(TowersBehindClus, *hcalTowers) / scRef->energy();
                             ^
RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:528:24: error: use of undeclared identifier 'egamma' [clang-diagnostic-error]
        invalidHcal = !egamma::hasActiveHcal(TowersBehindClus, *ctmaph, *hcalQuality, *hcalTopology);
                       ^
Suppressed 2579 warnings (2570 in non-user code, 8 NOLINT, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

guitargeek commented:
Seems the failing code-checks are related to the changes since the last IB. Is it possible to trigger the code-checks later again when the new IB is available?

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31892/19306

- This PR adds an extra 136KB to repository

cmsbuild commented:
A new Pull Request was created by  @guitargeek (Jonas Rembser) for master.

It involves the following packages:

DQM/Physics
EgammaAnalysis/ElectronTools
PhysicsTools/PatAlgos
RecoEcal/EgammaCoreTools
RecoEgamma/EgammaElectronProducers
RecoEgamma/EgammaHLTProducers
RecoEgamma/EgammaPhotonAlgos
RecoEgamma/EgammaPhotonProducers
RecoEgamma/EgammaTools
RecoEgamma/ElectronIdentification
RecoEgamma/PhotonIdentification
RecoHI/HiEgammaAlgos
RecoParticleFlow/PFClusterProducer
Validation/EcalClusters


@perrotta, @cmsbuild, @andrius-k, @kmaeshima, @fwyzard, @ErnestaP, @Martin-Grunewald, @jfernan2, @fioriNTU, @slava77, @jpata, @santocch can you please review it and eventually sign? Thanks.
@rappoccio, @gouskos, @felicepantaleo, @jainshilpi, @hatakeyamak, @emilbols, @argiro, @jazzitup, @thomreis, @varuns23, @seemasharmafnal, @mmarionncern, @JyothsnaKomaragiri, @ahinzmann, @smoortga, @lgray, @jdolen, @sobhatta, @cericeci, @ferencek, @mandrenguyen, @yetkinyilmaz, @Sam-Harper, @afiqaize, @rovere, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @dgulhan, @clelange, @HuguesBrun, @rchatter, @riga, @Fedespring, @kurtejung, @calderona, @yenjie, @lecriste, @cbernet, @gpetruc, @mariadalfonso, @andrzejnovak, @folguera this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 593a614caf05c32b0cdb2e0c12b9a89bbf47ab50
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b903f1/10200/summary.html
CMSSW: CMSSW_11_2_X_2020-10-21-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b903f1/10200/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544110
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2544081
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jfernan2 commented:
+1

jpata commented on RecoEcal/EgammaCoreTools/interface/EcalClusterLazyTools.h:
```diff
@@ -49,7 +88,9 @@ class EcalClusterLazyToolsBase {
   float SuperClusterSeedTime(const reco::SuperCluster &cluster);
   // get BasicClusterTime of the seed basic cluser of the supercluster
-  float SuperClusterTime(const reco::SuperCluster &cluster, const edm::Event &ev);
+  inline float SuperClusterTime(const reco::SuperCluster &cluster, const edm::Event &ev) {
```
this is the only change that's somehow not related to the ESGetToken migration. Did you check the impact on CPU time & code size?

guitargeek replied:
Hi, no I didn't check that. Why do you expect CPU time to increase? This is a very simple function, so I thought it might be more appropriate to define it in the header as `inline`. In theory this should actually save CPU time because there is not function call, but the main reason why I did it was to simplify the code by not separating declaration and definition of this function. And code size should also not increase I think, as the inlined function is just a function call itself.

jpata replied:
OK, thanks for the clarification. I don't expect any slowdown from inlining of course, I was just curious as to the motivation, which you now explained.

jpata replied:
(if anything, I was expecting a speedup, which I figured was the motivation for this change)

jpata reviewed (commented):
Thanks for this cleanup! Code review looks good, just a minor question on the inlined function `float SuperClusterTime(...)`. Perhaps it would be good just for safety to do a profiling check with the bot?

Apart from that, given that the PR looks technical, I see no reason to go forward with it.

jpata commented:
+reconstruction

- migrates EcalClusterLazyTools to use ESGetToken
- technical, no reco change is expected or observed
- private profiling reports no difference as expected: [new](http://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/31892/new) vs [orig](http://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/31892/orig)

santocch commented:
+1

Martin-Grunewald commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
